File: app/src/main/java/com/e3hi/geodrop/data/Drop.kt
package com.e3hi.geodrop.data

data class Drop(
    val id: String = "",
    val text: String = "",
    val description: String? = null,
    val lat: Double = 0.0,
    val lng: Double = 0.0,
    val createdBy: String = "",
    val createdAt: Long = 0L,
    val dropperUsername: String? = null,
    val isAnonymous: Boolean = false,
    val isDeleted: Boolean = false,
    val deletedAt: Long? = null,
    val decayDays: Int? = null,
    val groupCode: String? = null,
    val dropType: DropType = DropType.COMMUNITY,
    val businessId: String? = null,
    val businessName: String? = null,
    val contentType: DropContentType = DropContentType.TEXT,
    val mediaUrl: String? = null,
    val mediaMimeType: String? = null,
    val mediaData: String? = null,
    val mediaStoragePath: String? = null,
    val isNsfw: Boolean = false,
    val nsfwLabels: List<String> = emptyList(),
    val upvoteCount: Long = 0,
    val downvoteCount: Long = 0,
    val voteMap: Map<String, Long> = emptyMap(),
    val reportCount: Long = 0,
    val reportedBy: Map<String, Long> = emptyMap(),
    val redemptionCode: String? = null,
    val redemptionLimit: Int? = null,
    val redemptionCount: Int = 0,
    val redeemedBy: Map<String, Long> = emptyMap()
)

enum class DropType {
    COMMUNITY,
    RESTAURANT_COUPON,
    TOUR_STOP;

    companion object {
        fun fromRaw(raw: String?): DropType {
            if (raw.isNullOrBlank()) return COMMUNITY
            return entries.firstOrNull { it.name.equals(raw, ignoreCase = true) } ?: COMMUNITY
        }
    }
}

fun Drop.requiresRedemption(): Boolean {
    if (dropType != DropType.RESTAURANT_COUPON) return false
    return !redemptionCode.isNullOrBlank()
}

fun Drop.remainingRedemptions(): Int? {
    val limit = redemptionLimit ?: return null
    return (limit - redemptionCount).coerceAtLeast(0)
}

fun Drop.isBusinessDrop(): Boolean = dropType != DropType.COMMUNITY

fun Drop.isRedeemedBy(userId: String?): Boolean {
    if (userId.isNullOrBlank()) return false
    return redeemedBy.containsKey(userId)
}

private const val MILLIS_PER_DAY = 86_400_000L

fun Drop.decayAtMillis(): Long? {
    val days = decayDays?.takeIf { it > 0 } ?: return null
    val created = createdAt.takeIf { it > 0L } ?: return null
    return created + days * MILLIS_PER_DAY
}

fun Drop.isExpired(nowMillis: Long = System.currentTimeMillis()): Boolean {
    val expireAt = decayAtMillis() ?: return false
    return expireAt <= nowMillis
}

fun Drop.remainingDecayMillis(nowMillis: Long = System.currentTimeMillis()): Long? {
    val expireAt = decayAtMillis() ?: return null
    val remaining = expireAt - nowMillis
    return if (remaining > 0) remaining else 0L
}

enum class DropVoteType(val value: Int) {
    NONE(0),
    UPVOTE(1),
    DOWNVOTE(-1);

    companion object {
        fun fromRaw(raw: Long?): DropVoteType = when (raw?.toInt()) {
            1 -> UPVOTE
            -1 -> DOWNVOTE
            else -> NONE
        }
    }
}

enum class DropContentType {
    TEXT,
    PHOTO,
    AUDIO,
    VIDEO;

    companion object {
        fun fromRaw(value: String?): DropContentType {
            if (value.isNullOrBlank()) return TEXT
            return values().firstOrNull { it.name.equals(value, ignoreCase = true) } ?: TEXT
        }
    }
}

fun Drop.displayTitle(): String {
    val descriptionText = description.orEmpty()
    val baseTitle = when (dropType) {
        DropType.RESTAURANT_COUPON -> text.ifBlank { descriptionText.ifBlank { "Special offer" } }
        DropType.TOUR_STOP -> text.ifBlank { descriptionText.ifBlank { "Tour stop" } }
        DropType.COMMUNITY -> when (contentType) {
            DropContentType.TEXT -> text.ifBlank { descriptionText.ifBlank { "(No message)" } }
            DropContentType.PHOTO -> text.ifBlank { descriptionText }.ifBlank { "Photo drop" }
            DropContentType.AUDIO -> text.ifBlank { descriptionText }.ifBlank { "Audio drop" }
            DropContentType.VIDEO -> text.ifBlank { descriptionText }.ifBlank { "Video drop" }
        }
    }

    val username = dropperUsername?.trim()?.takeIf { it.isNotEmpty() }
    return if (!isAnonymous && !username.isNullOrEmpty()) {
        val handle = if (username.startsWith("@")) username else "@$username"
        "$handle dropped $baseTitle"
    } else {
        baseTitle
    }
}

fun Drop.mediaLabel(): String? = mediaUrl?.takeIf { it.isNotBlank() }

fun Drop.discoveryTitle(): String = when (dropType) {
    DropType.RESTAURANT_COUPON -> "Local business offer"
    DropType.TOUR_STOP -> "Guided tour stop"
    DropType.COMMUNITY -> when (contentType) {
        DropContentType.TEXT -> "Hidden note"
        DropContentType.PHOTO -> "Hidden photo drop"
        DropContentType.AUDIO -> "Hidden audio drop"
        DropContentType.VIDEO -> "Hidden video drop"
    }
}

fun Drop.discoveryDescription(): String {
    val descriptionText = description.orEmpty()
    return when (dropType) {
        DropType.RESTAURANT_COUPON -> descriptionText.ifBlank {
            "Unlock the details to redeem this business offer nearby."
        }
        DropType.TOUR_STOP -> descriptionText.ifBlank {
            "Pick up this stop to access the guided story or directions."
        }
        DropType.COMMUNITY -> when (contentType) {
            DropContentType.TEXT -> descriptionText.ifBlank { "Collect this drop to read the message inside." }
            DropContentType.PHOTO -> descriptionText.ifBlank { "Pick up this drop to reveal the photo." }
            DropContentType.AUDIO -> descriptionText.ifBlank { "Collect this drop to listen to the recording." }
            DropContentType.VIDEO -> descriptionText.ifBlank { "Collect this drop to watch the clip." }
        }
    }
}

fun Drop.voteScore(): Long = upvoteCount - downvoteCount

fun Drop.userVote(userId: String?): DropVoteType {
    if (userId.isNullOrBlank()) return DropVoteType.NONE
    return DropVoteType.fromRaw(voteMap[userId])
}

fun Drop.applyUserVote(userId: String, vote: DropVoteType): Drop {
    val previousVote = voteMap[userId]?.toInt() ?: 0
    val targetVote = vote.value
    if (previousVote == targetVote) return this

    var updatedUpvotes = upvoteCount
    var updatedDownvotes = downvoteCount
    val updatedMap = voteMap.toMutableMap()

    when (previousVote) {
        1 -> updatedUpvotes = (updatedUpvotes - 1).coerceAtLeast(0)
        -1 -> updatedDownvotes = (updatedDownvotes - 1).coerceAtLeast(0)
    }

    when (vote) {
        DropVoteType.UPVOTE -> {
            updatedUpvotes += 1
            updatedMap[userId] = 1L
        }
        DropVoteType.DOWNVOTE -> {
            updatedDownvotes += 1
            updatedMap[userId] = -1L
        }
        DropVoteType.NONE -> updatedMap.remove(userId)
    }

    return copy(
        upvoteCount = updatedUpvotes,
        downvoteCount = updatedDownvotes,
        voteMap = updatedMap
    )
}

File: ios/GeoDropIOS/Models/Drop.swift
import Foundation
import FirebaseFirestore

struct Drop: Identifiable, Equatable, Codable {
    var id: String
    var text: String
    var description: String?
    var latitude: Double
    var longitude: Double
    var createdBy: String
    var createdAt: Date
    var dropperUsername: String?
    var isAnonymous: Bool
    var isDeleted: Bool
    var deletedAt: Date?
    var decayDays: Int?
    var groupCode: String?
    var dropType: DropType
    var businessId: String?
    var businessName: String?
    var contentType: DropContentType
    var mediaURL: URL?
    var mediaMimeType: String?
    var mediaData: String?
    var mediaStoragePath: String?
    var isNsfw: Bool
    var nsfwLabels: [String]
    var likeCount: Int
    var likedBy: [String: Bool]
    var reportCount: Int
    var reportedBy: [String: TimeInterval]
    var redemptionCode: String?
    var redemptionLimit: Int?
    var redemptionCount: Int
    var redeemedBy: [String: TimeInterval]
    var collectedBy: [String: Bool]

    init(
        id: String = "",
        text: String = "",
        description: String? = nil,
        latitude: Double = 0,
        longitude: Double = 0,
        createdBy: String = "",
        createdAt: Date = Date(timeIntervalSince1970: 0),
        dropperUsername: String? = nil,
        isAnonymous: Bool = false,
        isDeleted: Bool = false,
        deletedAt: Date? = nil,
        decayDays: Int? = nil,
        groupCode: String? = nil,
        dropType: DropType = .community,
        businessId: String? = nil,
        businessName: String? = nil,
        contentType: DropContentType = .text,
        mediaURL: URL? = nil,
        mediaMimeType: String? = nil,
        mediaData: String? = nil,
        mediaStoragePath: String? = nil,
        isNsfw: Bool = false,
        nsfwLabels: [String] = [],
        likeCount: Int = 0,
        likedBy: [String: Bool] = [:],
        reportCount: Int = 0,
        reportedBy: [String: TimeInterval] = [:],
        redemptionCode: String? = nil,
        redemptionLimit: Int? = nil,
        redemptionCount: Int = 0,
        redeemedBy: [String: TimeInterval] = [:],
        collectedBy: [String: Bool] = [:]
    ) {
        self.id = id
        self.text = text
        self.description = description
        self.latitude = latitude
        self.longitude = longitude
        self.createdBy = createdBy
        self.createdAt = createdAt
        self.dropperUsername = dropperUsername
        self.isAnonymous = isAnonymous
        self.isDeleted = isDeleted
        self.deletedAt = deletedAt
        self.decayDays = decayDays
        self.groupCode = groupCode
        self.dropType = dropType
        self.businessId = businessId
        self.businessName = businessName
        self.contentType = contentType
        self.mediaURL = mediaURL
        self.mediaMimeType = mediaMimeType
        self.mediaData = mediaData
        self.mediaStoragePath = mediaStoragePath
        self.isNsfw = isNsfw
        self.nsfwLabels = nsfwLabels
        self.likeCount = likeCount
        self.likedBy = likedBy
        self.reportCount = reportCount
        self.reportedBy = reportedBy
        self.redemptionCode = redemptionCode
        self.redemptionLimit = redemptionLimit
        self.redemptionCount = redemptionCount
        self.redeemedBy = redeemedBy
        self.collectedBy = collectedBy
    }

    var coordinate: Coordinate {
        Coordinate(latitude: latitude, longitude: longitude)
    }

    var displayTitle: String {
        let baseTitle: String
        let desc = description?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        switch dropType {
        case .restaurantCoupon:
            baseTitle = text.nonEmpty ?? desc.nonEmpty ?? "Special offer"
        case .tourStop:
            baseTitle = text.nonEmpty ?? desc.nonEmpty ?? "Tour stop"
        case .community:
            switch contentType {
            case .text:
                baseTitle = text.nonEmpty ?? desc.nonEmpty ?? "(No message)"
            case .photo:
                baseTitle = text.nonEmpty ?? desc.nonEmpty ?? "Photo drop"
            case .audio:
                baseTitle = text.nonEmpty ?? desc.nonEmpty ?? "Audio drop"
            case .video:
                baseTitle = text.nonEmpty ?? desc.nonEmpty ?? "Video drop"
            }
        }

        guard !isAnonymous, let username = dropperUsername?.trimmingCharacters(in: .whitespaces), !username.isEmpty else {
            return baseTitle
        }
        let handle = username.hasPrefix("@") ? username : "@\(username)"
        return "\(handle) dropped \(baseTitle)"
    }

    var isExpired: Bool {
        guard let days = decayDays, days > 0 else { return false }
        let expiry = createdAt.addingTimeInterval(TimeInterval(days * 86_400))
        return expiry <= Date()
    }

    func requiresRedemption() -> Bool {
        dropType == .restaurantCoupon && !(redemptionCode?.isEmpty ?? true)
    }

    func isRedeemed(by userId: String?) -> Bool {
        guard let uid = userId, !uid.isEmpty else { return false }
        return redeemedBy[uid] != nil
    }

    func isLiked(by userId: String?) -> DropLikeStatus {
        guard let uid = userId, !uid.isEmpty else { return .none }
        return DropLikeStatus.from(raw: likedBy[uid])
    }
}

extension Drop {
    init?(document: DocumentSnapshot) {
        guard let data = document.data() else { return nil }
        
        let inlineData: String?
        if let string = data["mediaData"] as? String, !string.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            inlineData = string
        } else if let blob = data["mediaData"] as? Data, !blob.isEmpty {
            inlineData = blob.base64EncodedString()
        } else {
            inlineData = nil
        }
        self.init(
            id: document.documentID,
            text: data["text"] as? String ?? "",
            description: data["description"] as? String,
            latitude: data["lat"] as? Double ?? 0,
            longitude: data["lng"] as? Double ?? 0,
            createdBy: data["createdBy"] as? String ?? "",
            createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(timeIntervalSince1970: 0),
            dropperUsername: data["dropperUsername"] as? String ?? data["createdByUsername"] as? String,
            isAnonymous: data["isAnonymous"] as? Bool ?? false,
            isDeleted: data["isDeleted"] as? Bool ?? false,
            deletedAt: (data["deletedAt"] as? Timestamp)?.dateValue(),
            decayDays: data["decayDays"] as? Int ?? (data["decayDays"] as? NSNumber)?.intValue,
            groupCode: (data["groupCode"] as? String).flatMap { value in
                let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !trimmed.isEmpty else { return nil }
                return trimmed.caseInsensitiveCompare("PUBLIC") == .orderedSame ? nil : trimmed.uppercased()
            },
            dropType: DropType(rawValue: (data["dropType"] as? String ?? "").uppercased()) ?? .community,
            businessId: data["businessId"] as? String,
            businessName: data["businessName"] as? String,
            contentType: DropContentType(rawValue: (data["contentType"] as? String ?? "").uppercased()) ?? .text,
            mediaURL: (data["mediaUrl"] as? String).flatMap(URL.init(string:)),
            mediaMimeType: data["mediaMimeType"] as? String,
            mediaData: inlineData,
            mediaStoragePath: data["mediaStoragePath"] as? String,
            isNsfw: data["isNsfw"] as? Bool ?? false,
            nsfwLabels: data["nsfwLabels"] as? [String] ?? [],
            likeCount: (data["likeCount"] as? NSNumber)?.intValue ?? 0,
            likedBy: data["likedBy"] as? [String: Bool] ?? [:],
            reportCount: (data["reportCount"] as? NSNumber)?.intValue ?? 0,
            reportedBy: data["reportedBy"] as? [String: TimeInterval] ?? [:],
            redemptionCode: data["redemptionCode"] as? String,
            redemptionLimit: (data["redemptionLimit"] as? NSNumber)?.intValue,
            redemptionCount: (data["redemptionCount"] as? NSNumber)?.intValue ?? 0,
            redeemedBy: data["redeemedBy"] as? [String: TimeInterval] ?? [:],
            collectedBy: data["collectedBy"] as? [String: Bool] ?? [:]
        )
    }

    func toFirestoreData() -> [String: Any] {
        var data: [String: Any] = [
            "text": text,
            "lat": latitude,
            "lng": longitude,
            "createdBy": createdBy,
            "createdAt": Timestamp(date: createdAt),
            "isAnonymous": isAnonymous,
            "isDeleted": isDeleted,
            "dropType": dropType.rawValue,
            "contentType": contentType.rawValue,
            "isNsfw": isNsfw,
            "nsfwLabels": nsfwLabels,
            "likeCount": likeCount,
            "likedBy": likedBy,
            "reportCount": reportCount,
            "reportedBy": reportedBy,
            "redemptionCount": redemptionCount,
            "redeemedBy": redeemedBy,
            "collectedBy": collectedBy
        ]

        if let description { data["description"] = description }
        if let dropperUsername { data["dropperUsername"] = dropperUsername }
        if let decayDays { data["decayDays"] = decayDays }
        if let groupCode { data["groupCode"] = groupCode }
        if let businessId { data["businessId"] = businessId }
        if let businessName { data["businessName"] = businessName }
        if let mediaURL { data["mediaUrl"] = mediaURL.absoluteString }
        if let mediaMimeType { data["mediaMimeType"] = mediaMimeType }
        if let mediaStoragePath { data["mediaStoragePath"] = mediaStoragePath }
        if let mediaData { data["mediaData"] = mediaData }
        if let deletedAt { data["deletedAt"] = Timestamp(date: deletedAt) }
        if let redemptionCode { data["redemptionCode"] = redemptionCode }
        if let redemptionLimit { data["redemptionLimit"] = redemptionLimit }
        return data
    }
}

struct Coordinate: Hashable {
    var latitude: Double
    var longitude: Double
}

enum DropType: String, CaseIterable, Codable {
    case community = "COMMUNITY"
    case restaurantCoupon = "RESTAURANT_COUPON"
    case tourStop = "TOUR_STOP"
}

enum DropContentType: String, CaseIterable, Codable {
    case text = "TEXT"
    case photo = "PHOTO"
    case audio = "AUDIO"
    case video = "VIDEO"
}

enum DropLikeStatus: String, Codable {
    case none
    case liked

    static func from(raw: Any?) -> DropLikeStatus {
        switch raw {
        case let value as Bool:
            return value ? .liked : .none
        case let value as Int:
            return value != 0 ? .liked : .none
        case let value as NSNumber:
            return value.intValue != 0 ? .liked : .none
        case let value as String:
            let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            return ["liked", "like", "true", "1"].contains(normalized) ? .liked : .none
        default:
            return .none
        }
    }
}

private extension String {
    var nonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
